packages/tenant-enrollment-nodejs/src/pluginImplementation.ts

```diff
@@ -0,0 +1,192 @@
+
+export const getOverrideableTenantFunctionImplementation = (
+  config: SuperTokensPluginTenantEnrollmentPluginConfig,
+): OverrideableTenantFunctionImplementation => {
+  const implementation: OverrideableTenantFunctionImplementation = {
+    canUserJoinTenant: async function (tenantId, userIdentificationDetail) {
+      /**
+       * Check if the user can join the tenant based on the email domain
+       *
+       * @param email - The email of the user
+       * @param tenantId - The id of the tenant
+       * @returns true if the user can join the tenant, false otherwise
+       */
+
+      // Skip this for the public tenant
+      if (tenantId === "public") {
+        return {
+          canJoin: true,
+          reason: undefined,
+        };
+      }
+
+      // Check if the tenant is invite only in which case we
+      // can't allow the user to join
+      if (this.isTenantInviteOnly(tenantId)) {
+        return {
+          canJoin: false,
+          reason: this.getMessageForNoSignUpReason(NotAllowedToSignUpReason.INVITE_ONLY),
+        };
+      }
+
+      let canJoin = false;
+      let reason = undefined;
+      if (userIdentificationDetail.type === "email") {
+        canJoin = this.isMatchingEmailDomain(tenantId, userIdentificationDetail.email);
+        if (!canJoin) {
+          reason = this.getMessageForNoSignUpReason(NotAllowedToSignUpReason.EMAIL_DOMAIN_NOT_ALLOWED);
+        }
+      } else if (userIdentificationDetail.type === "thirdParty") {
+        canJoin = this.isApprovedIdPProvider(tenantId, userIdentificationDetail.thirdPartyId);
+        if (!canJoin) {
+          reason = this.getMessageForNoSignUpReason(NotAllowedToSignUpReason.IDP_NOT_ALLOWED);
+        }
+      } else if (userIdentificationDetail.type === "phoneNumber") {
+        // We don't really have a way to check anything for phones so we can
+        // allow signup.
+        return {
+          canJoin: true,
+        };
+      }
+
+      return {
+        canJoin,
+        reason,
+      };
+    },
+    handleTenantJoiningApproval: async function (
+      user: User,
+      tenantId: string,
+      associateLoginMethodDef: AssociateAllLoginMethodsOfUserWithTenant,
+      sendEmail: SendPluginEmail,
+      appUrl: string,
+      userContext: UserContext,
+      assignRoleToUserInTenant: AssignRoleToUserInTenant,
+    ) {
+      /**
+       * Handle the tenant joining functionality for the user.
+       *
+       * If the tenant requires approval, we will add a request for the
+       * user.
+       * If the tenant doesn't require approval, we will add them as a member
+       * right away.
+       *
+       * @param user - The user to handle the tenant joining for
+       * @param tenantId - The id of the tenant to handle the tenant joining for
+       * @param associateLoginMethodDef - The function to associate the login methods of the user with the tenant
+       */
+      // Skip this for the public tenant
+      if (tenantId === "public") {
```

Comment:
Why are we skipping this for public? Or at least this should be configurable.
Also, how does this interact with `allowPublicTenantAccess` in the other plugin?